### PR TITLE
[RFR] Automate/Explorer: Navmazing/Widgetastic/Collections conversion

### PR DESCRIPTION
**Pull request by mfalesni**

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit c4cd9ec1c55aa696316e8124e65a3a7019eba553:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 3370ee31d457f2465cf1fd9387f416065aa4b4da:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d125e4a3aad2d2e1d23a3133ff69768213bd30a4:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

File: cfme/automate/explorer/domain.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            for domain in domains:
+                if domain.name == name:
+                    checked_domains.append(domain)
+                    row[0].check()
+                    break
+
+            if set(domains) == set(checked_domains):
+                break
+
+        if set(domains) != set(checked_domains):
```
Comment (RonnyPfannschmidt) [inline cfme/automate/explorer/domain.py]: should we maybe use an assertion, so tests can see the difference

**Comment by dajoRH** [comment]
Lint report for commit f03344cc620cc638441573029366cbde9d11ee88:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 53c85033f52f62565368c0fd5da75dbb6978951f:

`cfme/automate/explorer/namespace.py`:
- :red_circle: Line 2:1: **F401** *'re' imported but unused*
- :red_circle: Line 3:1: **F401** *'cached_property' imported but unused*
- :red_circle: Line 6:1: **F401** *'Checkbox' imported but unused*

`cfme/tests/automate/test_namespace.py`:
- :heavy_exclamation_mark: Line 62:40: **W292** *no newline at end of file*

`widgetastic_manageiq.py`:
- :heavy_exclamation_mark: Line 427:99: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit c2b7e0dd7e12264774104a4ccb197bf5072e0a87:

`cfme/automate/explorer/namespace.py`:
- :red_circle: Line 2:1: **F401** *'re' imported but unused*
- :red_circle: Line 3:1: **F401** *'cached_property' imported but unused*
- :red_circle: Line 6:1: **F401** *'Checkbox' imported but unused*

`cfme/tests/automate/test_namespace.py`:
- :heavy_exclamation_mark: Line 62:40: **W292** *no newline at end of file*

`widgetastic_manageiq.py`:
- :heavy_exclamation_mark: Line 427:99: **W292** *no newline at end of file*

Please, rectify these issues :smirk: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d4b582c97904717beef1fede73eb71848ab3ef50:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit db8c6f729ffc0cd4cff3e98d3a597f66ea66d5c9:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 468ac9e5923e3c2a7e09d4aa8a6e5a0f8e7ab786:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ed2c85b1744c1acf8069b3f40e7f353b66fc73f7:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8b5bd091d5c42a1d423947d3a47525feb38a36d6:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 6697889b4b8edd825a3801dbc93f79d2cfe0b27d:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit ad06ca00017e5b263721d46e9705936540c22ff2:

Everything seems all right from lint perspective :smile: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 13ce57c2e9f9a28c6d7ad14d45aef7f04a48f93b:

`widgetastic_manageiq.py`:
- :red_circle: Line 4:1: **F401** *'WebDriverException' imported but unused*

Please, rectify these issues :smirk: .

Commit inspection report:

**No commit flaws detected**.
You have multiple commits in the PR. When you want your PR to be merged, please squash them. Do not squash them when working on the pull request as this allows to see the progress!

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/automate/explorer/namespace.py [hunk 2]
```diff
@@ -90,8 +90,13 @@ def create(self, name=None, description=None, cancel=False):
     def delete(self, *namespaces):
         all_page = navigate_to(self.parent, 'Details')
         namespaces = list(namespaces)
+        parents = set()
```
Comment (RonnyPfannschmidt) [inline cfme/automate/explorer/namespace.py]: `parents = {namespace.parent for namespace in namespaces}` 

set comprehensions are a win

**Comment by dajoRH** [comment]
Lint report for commit e1432ecb3e6c109fbc10905062f3e99c0e5aeea0:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8ca125c93cb5a86f39bec18d00d351bc670e8f77:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by mfalesni** [comment]
Hah, back in green :fist:

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4e419af64e2f42eeaa4f67cac557b444894ca94d:

`cfme/automate/explorer/domain.py`:
- :bangbang: Line 279:1: **E303** *too many blank lines (3)*

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4b68fa2185c996b61b80dfb85d2b8abb52f89a4b:

`cfme/automate/explorer/domain.py`:
- :bangbang: Line 279:1: **E303** *too many blank lines (3)*

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 4e76ce62d0a72a6e453fd516a2711ca331a78064:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 988cb910e714b0f5b12622da72229c4c69ae3bbb:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d7f41e984fa465319041093b332d06e564371790:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 52f4b781f46a72ad525ab95abe8eda955d14fce9:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit be00786557a7ff8268fa375c3b3d051a8209c971:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 8b57e00cd64fd8ce7690b5677bf18732a4b46ef3:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit f94f57d8e298826fc552cba4913fb66f6615e24d:

`cfme/tests/automate/test_class.py`:
- :red_circle: Line 6:1: **F401** *'error' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 2a9078dd85e5165eb746bccbc453a3242ce81b24:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit eac404d6c032e3194805af5bc9155c3d75d55d61:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit b34337967796828d51c96c5db8373cb0fdd2a786:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit d58370e1110c9289f207776df6b2566000605829:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 20faec18cf8fd7c4462753377e0096cdc0b2ef59:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 17d526dd0800473fd26154e0a8ce5dc65b7911ab:

Lint OK! :cake: :punch: :cookie: .


**No commit flaws detected**.
<mergeable/>

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 40e4aefae349d8734652d44fbc84c891db530a1d:

`cfme/automate/explorer/instance.py`:
- :bangbang: Line 63:1: **E302** *expected 2 blank lines, found 1*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 9d7adb170b5c345dd417e0bf4ec53dc6b4098d27:

`cfme/automate/explorer/domain.py`:
- :red_circle: Line 16:1: **F401** *'.common.Copiable' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit e71eb9b017fe6a0ed72acd94ac6d5bab414edf3f:

`cfme/automate/explorer/domain.py`:
- :red_circle: Line 16:1: **F401** *'.common.Copiable' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 83cf96b16cba98fee3b241bf5e203cb1349aec58:

`cfme/automate/explorer/domain.py`:
- :red_circle: Line 16:1: **F401** *'.common.Copiable' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 52ee9dcc64f601b0af21fc9fb3e7efaaf675dc0d:

`cfme/automate/explorer/domain.py`:
- :red_circle: Line 16:1: **F401** *'.common.Copiable' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mfalesni** [comment]
I have just squashed it in order to resolve the merge conflict.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mfalesni** [comment]
Damn rebase.

**Comment by mfalesni** [comment]
There *are* some errors, majority of it is the tests fail for non-automate reason, some automate tests fail but I want to take a stab on it later and there were some issues with the notifications overlaying in the method crud ...

File: cfme/automate/explorer/domain.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+                raise ItemNotFound('Domain named {} not found in the database'.format(self.name))
+        return self._db_id
+
+    @property
+    def db_object(self):
+        if self.db_id is None:
+            return None
+        table = self.appliance.db['miq_ae_namespaces']
+        try:
+            return self.appliance.db.session.query(table).filter(table.id == self.db_id)[0]
```
Comment (RonnyPfannschmidt) [inline cfme/automate/explorer/domain.py]: please use a `.first()` here instead of `try/except`

File: cfme/automate/explorer/domain.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        except IndexError:
+            return None
+
+    @property
```
Comment (RonnyPfannschmidt) [inline cfme/automate/explorer/domain.py]: this looks like a cached property

File: cfme/automate/explorer/domain.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        if self._enabled is None:
+            self._enabled = self.db_object.enabled
+        return self._enabled
+
+    @enabled.setter
+    def enabled(self, value):
+        self._enabled = value
+
+    @property
+    def locked(self):
```
Comment (RonnyPfannschmidt) [inline cfme/automate/explorer/domain.py]: same here

File: cfme/automate/explorer/domain.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+        self.description = description
+        self.enabled = enabled
+        self.locked = locked
+        self._db_id = None
+
+    def as_fill_value(self):
+        return self.name
+
+    @cached_property
+    def db_id(self):
```
Comment (RonnyPfannschmidt) [inline cfme/automate/explorer/domain.py]: this looks like this cached property was accidentally implemented like a property that caches would have been
is this a rebase artifact?

Comment (mfalesni) [inline cfme/automate/explorer/domain.py]: Possibly, I did a couple of rough rebases.

File: cfme/automate/explorer/domain.py (hunk 4 above)
Comment (mfalesni) [inline cfme/automate/explorer/domain.py]: Does cached property have setter? I need a specific behaviour when it is None.

Comment (mfalesni) [inline cfme/automate/explorer/domain.py]: Maybe use attrib to make the enabled and locked smart and pulling the correct values from the database when not specified.

**Comment by mfalesni** [comment]
I figured out the fields parametrized views locators weren't right. I hope this gives us a cleaner run.

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mfalesni** [comment]
Rebasing after I see the run's results

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

File: cfme/automate/explorer/domain.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+
+    def all(self):
+        table = self.appliance.db['miq_ae_namespaces']
+        query = self.appliance.db.session.query(
+            table.name, table.description, table.enabled, table.source)
+        query = query.filter(table.name != '$', table.parent_id == None)  # noqa
+        result = []
+        for name, description, enabled, source in query:
+            result.append(
+                Domain(
```
Comment (izapolsk) [inline cfme/automate/explorer/domain.py]: OPTIONAL: instantiate call would suit better here

**Comment by dajoRH** [comment]
Lint report for commit ec3a33078de37bf356a52964bb041297096724a8:

`cfme/automate/explorer/instance.py`:
- :bangbang: Line 10:22: **E999** *SyntaxError: invalid syntax*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 25f62116be99ef53fa6bac488a087fb07ab4b918:

`cfme/automate/explorer/instance.py`:
- :bangbang: Line 10:22: **E999** *SyntaxError: invalid syntax*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by mfalesni** [comment]
It is looking good now, the errors from automate are minimal.

**Comment by mfalesni** [comment]
Damn bugzilla error, rebased before the error was fixed, now a clean run is in progress (I hope)

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Pull request merged**